### PR TITLE
Private compare

### PR DESCRIPTION
**Pull request by bendecoste**
WIP but feel free to comment.  Need to figure out best practice for slice ops then we are pretty well complete

### CODE REVIEW

File: tensorflow_encrypted/protocol/pond.py [hunk 1]
```diff
@@ -460,6 +460,22 @@ def mul(self, x, y):
     def square(self, x):
         return self.dispatch('square', x)
 
+    @memoize
+    def mod(self, x, y):
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: what do you need mod for?

File: tensorflow_encrypted/protocol/securenn.py [hunk 2]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
     def select_share(self, x: PondTensor, y: PondTensor, bit: PondTensor) -> PondTensor:
         return x + bit * (y - x)
 
-    def private_compare(self, x, r, beta):
-        raise NotImplementedError
+    def _private_compare_beta0(self, input: PondPrivateTensor, rho: PondPublicTensor):
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: maybe rename `input` to `x`?

File: tensorflow_encrypted/protocol/securenn.py [hunk 3]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
...
+        with tf.device(self.server_0.device_name):
+            wa = 2 * rho * input
+            w = input - wa
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: `w = prot.bitwise_xor(input, rho)`

File: tensorflow_encrypted/protocol/securenn.py [hunk 4]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
...
+
+            w0_sum = tf.zeros(shape=w.shape)
+            # for i in range(bits-1, -1, -1):
+            # TODO -- subscripting
+            # sum = tf.reduce_sum(w[:, i + 1:], axis=1)
+            # w0_sum[:, i] = sum % p
+
+            c0 = (input * -1) + w0_sum
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: `c = r - x + 1 + w_sum`

File: tensorflow_encrypted/protocol/securenn.py [hunk 5]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
...
+
+        with tf.device(self.server_1.device_name):
+            wa = 2 * rho * input
+            wb = rho - wa
+            w = input + wb
+
+            w1_sum = tf.zeros(shape=w.shape)
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: maybe you can use [`tf.cumsum`](https://www.tensorflow.org/api_docs/python/tf/cumsum) for this

File: tensorflow_encrypted/protocol/securenn.py [hunk 6]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
...
+            # w1_sum[:, i] = sum % p
+
+            c1 = rho + (input * -1) + 1 + w0_sum
+
+        return c0, c1
+
+    def _private_compare_beta1(self, input: PondPrivateTensor, theta: PondPublicTensor):
+        with tf.device(self.server_0.device_name):
+            wa = 2 * theta * input
+            w = input - wa
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: `w = prot.bitwise_xor(input, theta)`

File: tensorflow_encrypted/protocol/securenn.py [hunk 7]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
...
+
+            w0_sum = tf.zeros(shape=w.shape)
+            # for i in range(bits-1, -1, -1):
+            # TODO -- subscripting
+            # sum = tf.reduce_sum(w[:, i + 1:], axis=1)
+            # w0_sum[:, i] = sum % p
+
+            c0 = (input * -1) + w0_sum
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: `c = x - t + 1 + w_sum`

File: tensorflow_encrypted/protocol/securenn.py [hunk 8]
```diff
@@ -87,8 +94,105 @@ def greater_equal(self, x: PondTensor, y: PondTensor) -> PondTensor:
...
+            # sum = tf.reduce_sum(w[:, i + 1:], axis=1)
+            # w1_sum[:, i] = sum % p
+
+            c1 = -theta + input + 1 + w0_sum
+
+        return c0, c1
+
+    def _private_compare_edge(self):
+        random_values_u = self.tensor_factory.Tensor.sample_random_tensor((bits,), modulus=p - 1) + 1
+        c0 = (random_values_u + 1) % p
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/securenn.py]: you shouldn't have to take `% p` manually like this, are you using PrimeTensor?

File: tensorflow_encrypted/protocol/pond.py (hunk 1 above)
Comment (mortendahl) [reply tensorflow_encrypted/protocol/pond.py]: I think I see why, but don't think you need it!

Comment (bendecoste) [reply tensorflow_encrypted/protocol/pond.py]: relic :)

**Pull request merged**